// sklearn/build/generated/model_selection/FixedThresholdClassifier.js
// src/generated/model_selection/FixedThresholdClassifier.ts
import crypto from "node:crypto";
var FixedThresholdClassifier = class {
  id;
  opts;
  _py;
  _isInitialized = false;
  _isDisposed = false;
  constructor(opts) {
    this.id = `FixedThresholdClassifier${crypto.randomUUID().split("-")[0]}`;
    this.opts = opts || {};
  }
  get py() {
    return this._py;
  }
  set py(pythonBridge) {
    this._py = pythonBridge;
  }
  /**
      Initializes the underlying Python resources.
  
      This instance is not usable until the `Promise` returned by `init()` resolves.
     */
  async init(py) {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (this._isInitialized) {
      return;
    }
    if (!py) {
      throw new Error(
        "FixedThresholdClassifier.init requires a PythonBridge instance"
      );
    }
    this._py = py;
    await this._py.ex`
import numpy as np
from sklearn.model_selection import FixedThresholdClassifier
try: bridgeFixedThresholdClassifier
except NameError: bridgeFixedThresholdClassifier = {}
`;
    await this._py.ex`ctor_FixedThresholdClassifier = {'estimator': ${this.opts["estimator"] ?? void 0}, 'threshold': ${this.opts["threshold"] ?? void 0}, 'pos_label': ${this.opts["pos_label"] ?? void 0}, 'response_method': ${this.opts["response_method"] ?? void 0}}

ctor_FixedThresholdClassifier = {k: v for k, v in ctor_FixedThresholdClassifier.items() if v is not None}`;
    await this._py.ex`bridgeFixedThresholdClassifier[${this.id}] = FixedThresholdClassifier(**ctor_FixedThresholdClassifier)`;
    this._isInitialized = true;
  }
  /**
      Disposes of the underlying Python resources.
  
      Once `dispose()` is called, the instance is no longer usable.
     */
  async dispose() {
    if (this._isDisposed) {
      return;
    }
    if (!this._isInitialized) {
      return;
    }
    await this._py.ex`del bridgeFixedThresholdClassifier[${this.id}]`;
    this._isDisposed = true;
  }
  /**
    Decision function for samples in `X` using the fitted estimator.
   */
  async decision_function(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FixedThresholdClassifier must call init() before decision_function()"
      );
    }
    await this._py.ex`pms_FixedThresholdClassifier_decision_function = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_FixedThresholdClassifier_decision_function = {k: v for k, v in pms_FixedThresholdClassifier_decision_function.items() if v is not None}`;
    await this._py.ex`res_FixedThresholdClassifier_decision_function = bridgeFixedThresholdClassifier[${this.id}].decision_function(**pms_FixedThresholdClassifier_decision_function)`;
    return this._py`res_FixedThresholdClassifier_decision_function.tolist() if hasattr(res_FixedThresholdClassifier_decision_function, 'tolist') else res_FixedThresholdClassifier_decision_function`;
  }
  /**
    Fit the classifier.
   */
  async fit(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error("FixedThresholdClassifier must call init() before fit()");
    }
    await this._py.ex`pms_FixedThresholdClassifier_fit = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'params': ${opts["params"] ?? void 0}}

pms_FixedThresholdClassifier_fit = {k: v for k, v in pms_FixedThresholdClassifier_fit.items() if v is not None}`;
    await this._py.ex`res_FixedThresholdClassifier_fit = bridgeFixedThresholdClassifier[${this.id}].fit(**pms_FixedThresholdClassifier_fit)`;
    return this._py`res_FixedThresholdClassifier_fit.tolist() if hasattr(res_FixedThresholdClassifier_fit, 'tolist') else res_FixedThresholdClassifier_fit`;
  }
  /**
      Get metadata routing of this object.
  
      Please check [User Guide](https://scikit-learn.org/stable/modules/generated/../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
     */
  async get_metadata_routing(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FixedThresholdClassifier must call init() before get_metadata_routing()"
      );
    }
    await this._py.ex`pms_FixedThresholdClassifier_get_metadata_routing = {'routing': ${opts["routing"] ?? void 0}}

pms_FixedThresholdClassifier_get_metadata_routing = {k: v for k, v in pms_FixedThresholdClassifier_get_metadata_routing.items() if v is not None}`;
    await this._py.ex`res_FixedThresholdClassifier_get_metadata_routing = bridgeFixedThresholdClassifier[${this.id}].get_metadata_routing(**pms_FixedThresholdClassifier_get_metadata_routing)`;
    return this._py`res_FixedThresholdClassifier_get_metadata_routing.tolist() if hasattr(res_FixedThresholdClassifier_get_metadata_routing, 'tolist') else res_FixedThresholdClassifier_get_metadata_routing`;
  }
  /**
    Predict the target of new samples.
   */
  async predict(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FixedThresholdClassifier must call init() before predict()"
      );
    }
    await this._py.ex`pms_FixedThresholdClassifier_predict = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_FixedThresholdClassifier_predict = {k: v for k, v in pms_FixedThresholdClassifier_predict.items() if v is not None}`;
    await this._py.ex`res_FixedThresholdClassifier_predict = bridgeFixedThresholdClassifier[${this.id}].predict(**pms_FixedThresholdClassifier_predict)`;
    return this._py`res_FixedThresholdClassifier_predict.tolist() if hasattr(res_FixedThresholdClassifier_predict, 'tolist') else res_FixedThresholdClassifier_predict`;
  }
  /**
    Predict logarithm class probabilities for `X` using the fitted estimator.
   */
  async predict_log_proba(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FixedThresholdClassifier must call init() before predict_log_proba()"
      );
    }
    await this._py.ex`pms_FixedThresholdClassifier_predict_log_proba = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_FixedThresholdClassifier_predict_log_proba = {k: v for k, v in pms_FixedThresholdClassifier_predict_log_proba.items() if v is not None}`;
    await this._py.ex`res_FixedThresholdClassifier_predict_log_proba = bridgeFixedThresholdClassifier[${this.id}].predict_log_proba(**pms_FixedThresholdClassifier_predict_log_proba)`;
    return this._py`res_FixedThresholdClassifier_predict_log_proba.tolist() if hasattr(res_FixedThresholdClassifier_predict_log_proba, 'tolist') else res_FixedThresholdClassifier_predict_log_proba`;
  }
  /**
    Predict class probabilities for `X` using the fitted estimator.
   */
  async predict_proba(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FixedThresholdClassifier must call init() before predict_proba()"
      );
    }
    await this._py.ex`pms_FixedThresholdClassifier_predict_proba = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None}

pms_FixedThresholdClassifier_predict_proba = {k: v for k, v in pms_FixedThresholdClassifier_predict_proba.items() if v is not None}`;
    await this._py.ex`res_FixedThresholdClassifier_predict_proba = bridgeFixedThresholdClassifier[${this.id}].predict_proba(**pms_FixedThresholdClassifier_predict_proba)`;
    return this._py`res_FixedThresholdClassifier_predict_proba.tolist() if hasattr(res_FixedThresholdClassifier_predict_proba, 'tolist') else res_FixedThresholdClassifier_predict_proba`;
  }
  /**
      Return the mean accuracy on the given test data and labels.
  
      In multi-label classification, this is the subset accuracy which is a harsh metric since you require for each sample that each label set be correctly predicted.
     */
  async score(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FixedThresholdClassifier must call init() before score()"
      );
    }
    await this._py.ex`pms_FixedThresholdClassifier_score = {'X': np.array(${opts["X"] ?? void 0}) if ${opts["X"] !== void 0} else None, 'y': np.array(${opts["y"] ?? void 0}) if ${opts["y"] !== void 0} else None, 'sample_weight': np.array(${opts["sample_weight"] ?? void 0}) if ${opts["sample_weight"] !== void 0} else None}

pms_FixedThresholdClassifier_score = {k: v for k, v in pms_FixedThresholdClassifier_score.items() if v is not None}`;
    await this._py.ex`res_FixedThresholdClassifier_score = bridgeFixedThresholdClassifier[${this.id}].score(**pms_FixedThresholdClassifier_score)`;
    return this._py`res_FixedThresholdClassifier_score.tolist() if hasattr(res_FixedThresholdClassifier_score, 'tolist') else res_FixedThresholdClassifier_score`;
  }
  /**
      Request metadata passed to the `score` method.
  
      Note that this method is only relevant if `enable_metadata_routing=True` (see [`sklearn.set_config`](https://scikit-learn.org/stable/modules/generated/sklearn.set_config.html#sklearn.set_config "sklearn.set_config")). Please see [User Guide](https://scikit-learn.org/stable/modules/generated/../../metadata_routing.html#metadata-routing) on how the routing mechanism works.
  
      The options for each parameter are:
     */
  async set_score_request(opts) {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FixedThresholdClassifier must call init() before set_score_request()"
      );
    }
    await this._py.ex`pms_FixedThresholdClassifier_set_score_request = {'sample_weight': ${opts["sample_weight"] ?? void 0}}

pms_FixedThresholdClassifier_set_score_request = {k: v for k, v in pms_FixedThresholdClassifier_set_score_request.items() if v is not None}`;
    await this._py.ex`res_FixedThresholdClassifier_set_score_request = bridgeFixedThresholdClassifier[${this.id}].set_score_request(**pms_FixedThresholdClassifier_set_score_request)`;
    return this._py`res_FixedThresholdClassifier_set_score_request.tolist() if hasattr(res_FixedThresholdClassifier_set_score_request, 'tolist') else res_FixedThresholdClassifier_set_score_request`;
  }
  /**
    The fitted classifier used when predicting.
   */
  get estimator_() {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FixedThresholdClassifier must call init() before accessing estimator_"
      );
    }
    return (async () => {
      await this._py.ex`attr_FixedThresholdClassifier_estimator_ = bridgeFixedThresholdClassifier[${this.id}].estimator_`;
      return this._py`attr_FixedThresholdClassifier_estimator_.tolist() if hasattr(attr_FixedThresholdClassifier_estimator_, 'tolist') else attr_FixedThresholdClassifier_estimator_`;
    })();
  }
  /**
    Number of features seen during [fit](https://scikit-learn.org/stable/modules/generated/../../glossary.html#term-fit). Only defined if the underlying estimator exposes such an attribute when fit.
   */
  get n_features_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FixedThresholdClassifier must call init() before accessing n_features_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_FixedThresholdClassifier_n_features_in_ = bridgeFixedThresholdClassifier[${this.id}].n_features_in_`;
      return this._py`attr_FixedThresholdClassifier_n_features_in_.tolist() if hasattr(attr_FixedThresholdClassifier_n_features_in_, 'tolist') else attr_FixedThresholdClassifier_n_features_in_`;
    })();
  }
  /**
    Names of features seen during [fit](https://scikit-learn.org/stable/modules/generated/../../glossary.html#term-fit). Only defined if the underlying estimator exposes such an attribute when fit.
   */
  get feature_names_in_() {
    if (this._isDisposed) {
      throw new Error(
        "This FixedThresholdClassifier instance has already been disposed"
      );
    }
    if (!this._isInitialized) {
      throw new Error(
        "FixedThresholdClassifier must call init() before accessing feature_names_in_"
      );
    }
    return (async () => {
      await this._py.ex`attr_FixedThresholdClassifier_feature_names_in_ = bridgeFixedThresholdClassifier[${this.id}].feature_names_in_`;
      return this._py`attr_FixedThresholdClassifier_feature_names_in_.tolist() if hasattr(attr_FixedThresholdClassifier_feature_names_in_, 'tolist') else attr_FixedThresholdClassifier_feature_names_in_`;
    })();
  }
};
export {
  FixedThresholdClassifier
};
